domain/user/state/state.go

```diff
@@ -0,0 +1,395 @@
+	}
+	if removed {
+		return fmt.Errorf("cannot set password hash for removed user %q", uuid)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removeActivationKey(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setPasswordHash(ctx, tx, uuid, passwordHash, salt)
+	})
+}
+
+func addUser(ctx context.Context, tx *sql.Tx, uuid user.UUID, user user.User, creatorUuid user.UUID) error {
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	_, err := tx.ExecContext(ctx, addUserQuery,
+		uuid.String(),
+		user.Name,
+		user.DisplayName,
+		creatorUuid.String(),
+		time.Now(),
+	)
+	if databaseutils.IsErrConstraintUnique(err) {
+		return usererrors.AlreadyExists
+	} else if databaseutils.IsErrConstraintForeignKey(err) {
+		return usererrors.UserCreatorUuidNotFound
+	} else if err != nil {
+		return fmt.Errorf("cannot insert user %q: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+func setPasswordHash(ctx context.Context, tx *sql.Tx, uuid user.UUID, passwordHash string, salt []byte) error {
+	defineUserAuthenticationQuery := `
+INSERT INTO user_authentication (user_uuid, disabled) 
+VALUES (?, ?) 
+ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid, disabled = excluded.disabled
+`
+
+	_, err := tx.ExecContext(ctx, defineUserAuthenticationQuery, uuid, false)
+	if err != nil {
+		return fmt.Errorf("cannot set authentification for user %q: %w", uuid, err)
+	}
+
+	setPasswordHashQuery := `
+INSERT INTO user_password (user_uuid, password_hash, password_salt) 
+VALUES (?, ?, ?) 
+ON CONFLICT(user_uuid) DO UPDATE SET password_hash = excluded.password_hash, password_salt = excluded.password_salt
+`
+
+	_, err = tx.ExecContext(ctx, setPasswordHashQuery, uuid.String(), passwordHash, salt)
+	if err != nil {
+		return fmt.Errorf("cannot set password hash for user %q: %w", uuid, err)
+	}
+
+	return nil
+}
+
+func removePasswordHash(ctx context.Context, tx *sql.Tx, uuid user.UUID) error {
+	removePasswordHashQuery := `
+DELETE FROM user_password
+WHERE user_uuid = ?
+`
+
+	_, err := tx.ExecContext(ctx, removePasswordHashQuery, uuid.String())
+	if err != nil {
+		return fmt.Errorf("cannot remove password hash for user %q: %w", uuid, err)
+	}
+
+	return nil
+}
+
+func removeActivationKey(ctx context.Context, tx *sql.Tx, uuid user.UUID) error {
+	removeActivationKeyQuery := `
+DELETE FROM user_activation_key
+WHERE user_uuid = ?
+`
+
+	_, err := tx.ExecContext(ctx, removeActivationKeyQuery, uuid.String())
+	if err != nil {
+		return fmt.Errorf("cannot remove activation key for user %q: %w", uuid, err)
+	}
+
+	return nil
+}
+
+func getUser(ctx context.Context, tx *sql.Tx, uuid user.UUID) (user.User, error) {
+	getUserQuery := `
+SELECT name, display_name, created_by_uuid, created_at
+FROM user
+WHERE uuid = ? AND removed = false
```

Comment:
See comments in services layer but I think now that we are requesting users based on UUID in this context we don't actually care if they're removed?

Reply:
agreed